src/audio_unit/looper.rs

```diff
@@ -0,0 +1,137 @@
+
+        (
+            Self {
+                messages: receiver,
+                buffer,
+                state: Off,
+            },
+            sender,
+        )
+    }
+
+    fn process_messages(&mut self) {
+        let messages: Vec<_> = self.messages.try_iter().collect();
+        for message in messages {
+            self.process_message(message);
+        }
+    }
+
+    fn process_message(&mut self, _: Message) {
+        self.state = match self.state {
+            Off => Recording { position: 0 },
+            Recording { position } => Playing {
+                position: 0,
+                total: position,
+            },
+            Playing {
+                position: _,
+                total: _,
+            } => {
+                util::zero_slice(&mut self.buffer);
+                Off
+            }
+        };
+
+        println!("looper: {:?}", self.state);
+    }
+
+    fn process_samples(&mut self, input: &[f32], output: &mut [f32]) {
+        match self.state {
+            Off => {
+                util::zero_slice(output);
+            }
+            Recording { position } => {
+                let next_position = position + input.len();
+
+                if next_position <= self.buffer.len() {
+                    self.buffer[position..next_position].copy_from_slice(input);
+
+                    if next_position == self.buffer.len() {
+                        println!("looper: out of space in the buffer. switching to playback");
+
+                        self.state = Playing {
+                            position: 0,
+                            total: next_position,
+                        };
+                    } else {
+                        self.state = Recording {
+                            position: next_position,
+                        };
+                    }
+                } else {
+                    self.process_samples_wrap_around(position, self.buffer.len(), input, output);
+                }
+            }
+            Playing { position, total } => {
+                let next_position = position + output.len();
+
+                if next_position <= total {
+                    output.copy_from_slice(&self.buffer[position..next_position]);
+
+                    self.state = Playing {
+                        position: next_position % total,
+                        total,
+                    }
+                } else {
+                    self.process_samples_wrap_around(position, total, input, output);
+                }
+            }
+        };
+    }
+
+    fn process_samples_wrap_around(
+        &mut self,
+        position: usize,
+        total: usize,
+        input: &[f32],
+        output: &mut [f32],
+    ) {
+        let split_position = total - position;
+
+        // play/record to end of buffer
+        self.process_samples(&input[0..split_position], &mut output[0..split_position]);
+
+        // ... then wrap around
+        self.process_samples(
```

Comment:
Ok so because `process_samples()` automatically switches to `Playing` once it (exactly) hits the end of its buffer size, this will now be in `Playing` mode (so result in playback of the first however many samples in the buffer) regardless of whether we entered `process_samples_wrap_around()` in `Recording` or `Playing` mode?

Reply:
Right